ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/dns/DNSDaemonTest.java

```diff
@@ -0,0 +1,121 @@
+import io.vertx.core.ThreadingModel;
+import io.vertx.core.Vertx;
+import io.vertx.junit5.Checkpoint;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(VertxExtension.class)
+class DNSDaemonTest {
+  private static final String holeskyEnr =
+      "enrtree://AKA3AM6LPBYEUDMVNU3BSVQJ5AD45Y7YPOHJLEF6W26QOE4VTUDPE@all.holesky.ethdisco.net";
+  private static MockDNSServer mockDNSServer;
+  private DNSDaemon dnsDaemon;
+
+  @BeforeAll
+  static void setup() throws IOException {
+    Security.addProvider(new BouncyCastleProvider());
+    mockDNSServer = new MockDNSServer();
+    mockDNSServer.start();
+  }
+
+  @AfterAll
+  static void tearDown() {
+    mockDNSServer.stop();
+  }
+
+  @Test
+  @DisplayName("Test DNS Daemon with a mock DNS server")
+  void testDNSDaemon(final Vertx vertx, final VertxTestContext testContext)
+      throws InterruptedException {
+    final Checkpoint checkpoint = testContext.checkpoint();
+    dnsDaemon =
+        new DNSDaemon(
+            holeskyEnr,
+            (seq, records) -> checkpoint.flag(),
+            0,
+            0,
+            0,
+            "localhost:" + mockDNSServer.port());
+
+    final DeploymentOptions options =
+        new DeploymentOptions().setThreadingModel(ThreadingModel.WORKER).setWorkerPoolSize(1);
+    vertx.deployVerticle(dnsDaemon, options);
+  }
+
+  @Test
+  @DisplayName("Test DNS Daemon with periodic lookup to a mock DNS server")
+  void testDNSDaemonPeriodic(final Vertx vertx, final VertxTestContext testContext)
+      throws InterruptedException {
+    // checkpoint should be flagged twice
+    final Checkpoint checkpoint = testContext.checkpoint(2);
+    final AtomicInteger pass = new AtomicInteger(0);
+    dnsDaemon =
+        new DNSDaemon(
+            holeskyEnr,
+            (seq, records) -> {
+              switch (pass.incrementAndGet()) {
+                case 1:
+                  testContext.verify(
+                      () -> {
+                        assertThat(seq).isEqualTo(932);
+                        assertThat(records).hasSize(115);
+                      });
+                  break;
+                case 2:
+                  testContext.verify(
+                      () -> {
+                        assertThat(seq).isEqualTo(932);
+                        assertThat(records).isEmpty();
+                      });
+                  break;
+                default:
+                  testContext.failNow("Third pass is not expected");
+              }
+              checkpoint.flag();
+            },
+            0,
+            100, // starts the lookup after 100 ms
```

Comment:
We could save some time and reduce these to 1, and 50?

Reply:
done.